### PR TITLE
Support Substrait's VirtualTables

### LINKED ISSUE #10530: Support Substrait VirtualTables
### Is your feature request related to a problem or challenge?

I'm working on Spark -> Substrait -> DataFusion conversion, and as part of it we sometimes, especially during testing,use Substrait's VirtualTables (ie tables with data built-in rather than read from a source). 

Currently however reading those VirtualTables into DataFusion fails with `NotImplemented("Only NamedTable reads are supported")`

I'm working on a PR to add the support.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Blizzara**
## Which issue does this PR close?

Closes #10530

This (well, now together with the earlier PRs for chunks of this) is my first (but hopefully not last!) contribution to DataFusion, and I'm pretty new with Rust as well - so all feedback on both style and details is very much welcome!

## What changes are included in this PR?

Adds support for Substrait's VirtualTables, ie. tables with data baked-in into the Substrait plan instead of being read from a source. 

Adds conversion in both ways (Substrait -> DataFusion and DataFusion -> Substrait)

Adds support for both empty VirtualTables (mapped into LogicalPlan::EmptyTable) and those containing values (mapped into LogicalPlan::Values).

## Are these changes tested?

Tested through round-trip tests, similar to other Substrait support.

## Are there any user-facing changes?
 

- More support through Substrait

### CODE REVIEW

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 1]
```diff
@@ -1137,7 +1220,8 @@ fn from_substrait_type(dt: &substrait::proto::Type) -> Result<DataType> {
                     from_substrait_type(list.r#type.as_ref().ok_or_else(|| {
                         substrait_datafusion_err!("List type must have inner type")
                     })?)?;
-                let field = Arc::new(Field::new("list_item", inner_type, true));
+                // "item" as the field name aligns with Arrow's default, see e.g. array_into_list_array
+                let field = Arc::new(Field::new("item", inner_type, true));
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:1224]: this is in theory a breaking change I guess, but I don't know if there's any reason for anyone to be referring to the item field name

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 2]
```diff
@@ -1277,6 +1407,56 @@ pub(crate) fn from_substrait_literal(lit: &Literal) -> Result<ScalarValue> {
                 s,
             )
         }
+        Some(LiteralType::List(l)) => {
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:1410]: I don't necessarily need these for this PR, so if it's preferrable I'm happy to split them into separate PRs - I just added them while coding to confirm that lists and structs inside lists are handled correctly.

File: datafusion/substrait/src/logical_plan/producer.rs [hunk 3]
```diff
@@ -165,6 +168,53 @@ pub fn to_substrait_rel(
                 }))),
             }))
         }
+        LogicalPlan::Values(v) => {
+            let names = v
+                .schema
+                .fields()
+                .iter()
+                .map(|f| f.name().to_owned())
+                // todo: collect also nested names
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/producer.rs:177]: note for self: todo

File: datafusion/substrait/src/logical_plan/producer.rs [hunk 4]
```diff
@@ -165,6 +168,53 @@ pub fn to_substrait_rel(
...
+            };
+            let values = v
+                .values
+                .iter()
+                .map(|row| {
+                    let fields = row
+                        .iter()
+                        .map(|v| match v {
+                            Expr::Literal(sv) => to_substrait_literal(sv).unwrap(),
+                            _ => panic!(),
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/producer.rs:197]: note for self: better error

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs [hunk 5]
```diff
@@ -607,6 +608,15 @@ async fn qualified_catalog_schema_table_reference() -> Result<()> {
     roundtrip("SELECT a,b,c,d,e FROM datafusion.public.data;").await
 }
 
+#[tokio::test]
+async fn roundtrip_local_relation() -> Result<()> {
+    let ctx = create_context().await?;
+    let row1 = vec![lit(1), lit("a")];
+    let row2 = vec![lit(2), lit("b")];
+    let plan = LogicalPlanBuilder::values(vec![row1, row2])?.build()?;
```
Comment (Blizzara) [inline datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:616]: I didn't find `VALUES` in the DataFusion SQL - is it not implemented yet?

**Comment by alamb** [comment]
Thanks @Blizzara  -- I started the CI on this PR

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs (hunk 5 above)
Comment (jonahgao) [reply datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:646]: It is implemented, and you can use it like this.
```rust
roundtrip("VALUES (1, 'a', [[1,2,3], []], STRUCT(true, 1))").await
```

**Comment by Blizzara** [comment]
@jonahgao @alamb I think this is ready by me :)

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 6]
```diff
@@ -1277,7 +1404,84 @@ pub(crate) fn from_substrait_literal(lit: &Literal) -> Result<ScalarValue> {
                 s,
             )
         }
-        Some(LiteralType::Null(ntype)) => from_substrait_null(ntype)?,
+        Some(LiteralType::List(l)) => {
+            let element_dt: Option<&DataType> = match dt {
+                Some(DataType::List(l)) => Ok(Some(l.data_type())),
```
Comment (jonahgao) [inline datafusion/substrait/src/logical_plan/consumer.rs:1409]: I am wondering if `dt` is redundant here and can be removed. We still need to check whether it is a `LargeList`, just as we check `type_variation_reference` below.

Is it possible to only extract `element_dt` from `elements[0]`?  I guess that `elements` are not empty because substrait has a special type called EmptyList.
```rust
 pub struct List {
            /// A homogeneously-typed list of one or more expressions that form the
            /// list entries. To specify an empty list, use Literal.empty_list
            /// (otherwise type information would be missing).
            #[prost(message, repeated, tag="1")]
            pub values: ::prost::alloc::vec::Vec<super::super::Expression>,
        }
```

File: datafusion/substrait/src/logical_plan/consumer.rs (hunk 2 above)
Comment (jonahgao) [reply datafusion/substrait/src/logical_plan/consumer.rs:1407]: If it is convenient, using separate PRs would make the review process easier. Perhaps one PR could focus on the virtual table, while the other two could address the list and struct separately.

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 7]
```diff
@@ -505,7 +507,35 @@ pub async fn from_substrait_rel(
...
             }
-            _ => not_impl_err!("Only NamedTable reads are supported"),
+            Some(ReadType::VirtualTable(vt)) => {
+                let base_schema = read
+                    .base_schema
+                    .clone()
+                    .expect("expected schema to exist for virtual table");
+
+                let fields = from_substrait_named_struct(&base_schema)?;
+                let schema = DFSchemaRef::new(DFSchema::try_from(Schema::new(fields))?);
```
Comment (jonahgao) [inline datafusion/substrait/src/logical_plan/consumer.rs:517]: I suspect that the schema here may not be necessary, perhaps it can be restored through [LogicalPlanBuilder::values](https://github.com/apache/datafusion/blob/f80794779ef251528ee983e22ec03c158deb42a6/datafusion/expr/src/logical_plan/builder.rs#L168).

Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs:517]: It probably could, except for the column names - LogicalPlanBuilder::values names everything to just `column1` etc. Do you see some benefit to using that (and then handling the column names somehow) instead of passing in the schema explicitly?

File: datafusion/substrait/src/logical_plan/consumer.rs (hunk 6 above)
Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs:1409]: Yeah, we already do that (look at the first element) if `dt` is not given. I think the reason I included `dt` was because Substrait types don't include the field names for structs, so not using `dt` would mean using something like `names` and `name_idx` instead (like in from_substrait_type_with_names).

Comment (jonahgao) [reply datafusion/substrait/src/logical_plan/consumer.rs:1409]: Thank you for your explanation @Blizzara . My concern is that using two types for List simultaneously is a bit verbose unless it is necessary.

I tested the following two cases: the first one panicked, and the second one also failed.
```rust

#[tokio::test]
async fn roundtrip_list_iteral() -> Result<()> {
    roundtrip("SELECT [] from data").await
}

#[tokio::test]
async fn roundtrip_struct_iteral() -> Result<()> {
    roundtrip("select struct(1 as name0, 3.14 as name1, 'e', true as name3) from data")
        .await
}
``` 
```sh
---- cases::roundtrip_logical_plan::roundtrip_list_iteral stdout ----
thread 'cases::roundtrip_logical_plan::roundtrip_list_iteral' panicked at consumer.rs:1425:41:
index out of bounds: the len is 0 but the index is 0
```
I think lists and structs are more complex than the virtual table, so it's better for them to have separate PRs.

File: datafusion/substrait/src/logical_plan/consumer.rs (hunk 7 above)
Comment (jonahgao) [reply datafusion/substrait/src/logical_plan/consumer.rs:517]: I once wanted to remove the schema, but we need to use it to pass the struct's names, so the current approach should be fine.

File: datafusion/substrait/src/logical_plan/consumer.rs (hunk 6 above)
Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs:1409]: This was because I wasn't handling empty lists properly (producer created Lists rather than EmptyList) - that was fixed in the Lists PR. Thanks for catching!

File: datafusion/substrait/src/logical_plan/consumer.rs (hunk 2 above)
Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs]: Split those into https://github.com/apache/datafusion/pull/10615, https://github.com/apache/datafusion/pull/10622, https://github.com/apache/datafusion/pull/10640

**Comment by Blizzara** [comment]
@jonahgao @alamb this last one is ready now too :)

File: datafusion/substrait/src/logical_plan/producer.rs [hunk 8]
```diff
@@ -512,6 +572,62 @@ pub fn to_substrait_rel(
...
+        match dtype {
+            DataType::Struct(fields) => {
+                let mut names = Vec::new();
+                for field in fields {
+                    names.push(field.name().to_string());
+                    names.extend(names_dfs(field.data_type())?);
+                }
+                Ok(names)
+            }
+            DataType::List(l) => names_dfs(l.data_type()),
```
Comment (jonahgao) [inline datafusion/substrait/src/logical_plan/producer.rs:589]: I think we also need to consider `DataType::LargeList` here because `to_substrait_type` supports it.

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs [hunk 9]
```diff
@@ -685,6 +685,19 @@ async fn roundtrip_literal_struct() -> Result<()> {
...
+async fn roundtrip_values() -> Result<()> {
+    assert_expected_plan(
+        "VALUES (1, 'a', [[-213.1, NULL, 5.5, 2.0, 1.0], []], STRUCT(true, 1, CAST(NULL AS STRING))), (NULL, NULL, NULL, NULL)",
+        "Values: (Int64(1), Utf8(\"a\"), List([[-213.1, , 5.5, 2.0, 1.0], []]), Struct({c0:true,c1:1,c2:})), (Int64(NULL), Utf8(NULL), List(), Struct({c0:,c1:,c2:}))")
+        .await
+}
+
+#[tokio::test]
+async fn roundtrip_empty_relation() -> Result<()> {
+    roundtrip("SELECT * FROM (VALUES (1)) LIMIT 0").await
```
Comment (jonahgao) [inline datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:698]: I've slightly modified this test and printed out the schema on the consumer side, but it seems to be incorrect.
```rust
#[tokio::test]
async fn roundtrip_empty_relation() -> Result<()> {
    roundtrip("SELECT * FROM (VALUES ([STRUCT(1 as a)], 2)) LIMIT 0").await
}
```
```sh
plan2 schema: DFSchema { inner: Schema { fields: [
Field { name: "column1", data_type: List(Field { name: "item", data_type: Struct([Field { name: "c0", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }]), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, 
# Its name should be "column2"
Field { name: "a", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, 
field_qualifiers: [None, None], functional_dependencies: FunctionalDependencies { deps: [] } }
```

Comment (jonahgao) [reply datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:698]: We might need a test to verify the schemas. For this test, the plan str on both sides would just be `EmptyRelation`.

Comment (Blizzara) [reply datafusion/substrait/tests/cases/roundtrip_logical_plan.rs]: Good catch! Fixed in https://github.com/apache/datafusion/pull/10531/commits/24d07c9d92d90f6f3287caca4b70e0f6e2d971f6 and added schema verification to most tests

File: datafusion/substrait/src/logical_plan/producer.rs (hunk 8 above)
Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/producer.rs:589]: Thanks, fixed https://github.com/apache/datafusion/pull/10531/commits/9c83ba469ece074a9611f4db871f2e0044652605

Comment (jonahgao) [reply datafusion/substrait/src/logical_plan/producer.rs:596]: `arrow_cast` is a solution .
```sql
VALUES(arrow_cast([1,2], 'LargeList(Int64)'));
```

Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/producer.rs:596]: Nice! I added that in https://github.com/apache/datafusion/pull/10531/commits/4b8babe5353cff2740c1211cff6973dc83264098, though it seems that arrow_cast doesn't [yet support structs](https://github.com/apache/datafusion/blob/2e6beb434d45693bc8c7ba636b0407c26c04fc37/datafusion/functions/src/core/arrow_cast.rs#L194), so the test still wouldn't have caught this. Unless you can think of some workaround? I'm not familiar at all with arrow_cast and the like 😅

Comment (jonahgao) [reply datafusion/substrait/src/logical_plan/producer.rs:596]: I think so too, let's keep it as a TODO.

**Review by jonahgao — APPROVED** [review]
Looks good to me👍. Thank you for your contribution @Blizzara .
